packages/itwin/tree-widget/src/components/trees/models-tree/internal/FilteredTree.ts

```diff
@@ -0,0 +1,255 @@
+        break;
+      }
+
+      const identifier = normalizedPath[i];
+      if (!HierarchyNodeIdentifier.isInstanceNodeIdentifier(identifier)) {
+        break;
+      }
+
+      const currentNode: FilteredTreeNode | undefined = parentNode.children?.get(identifier.id);
+      if (currentNode !== undefined) {
+        parentNode = currentNode;
+        continue;
+      }
+
+      const type = await getType(imodelAccess, identifier.className);
+      const newNode: FilteredTreeNode = createFilteredTreeNode({
+        type,
+        id: identifier.id,
+        isFilterTarget: i === normalizedPath.length - 1,
+        parent: parentNode as FilteredTreeNode,
+      });
+      (parentNode.children ??= new Map()).set(identifier.id, newNode);
+      parentNode = newNode;
+    }
+  }
+
+  return {
+    getFilterTargets: (node: HierarchyNode) => getFilterTargets(root, node),
+  };
+}
+
+function getFilterTargets(root: FilteredTreeRootNode, node: HierarchyNode) {
+  let lookupParents: Array<{ children?: Map<Id64String, FilteredTreeNode> }> = [root];
+  const filterTargets: FilterTargets = {};
+
+  const nodeKey = node.key;
+  if (!HierarchyNodeKey.isInstances(nodeKey)) {
+    return filterTargets;
+  }
+
+  // find the filtered parent nodes of the `node`
+  for (const parentKey of node.parentKeys) {
+    if (!HierarchyNodeKey.isInstances(parentKey)) {
+      continue;
+    }
+
+    // tree node might be merged from multiple instances. As filtered tree stores only one instance per node, we need to find all matching nodes
+    // and use them when checking for matching node in one level deeper.
+    const parentNodes = findMatchingFilteredNodes(lookupParents, parentKey.instanceKeys);
+    if (parentNodes.length === 0) {
+      return filterTargets;
+    }
+    lookupParents = parentNodes;
+  }
+
+  // find filtered nodes that match the `node`
+  const filteredNodes = findMatchingFilteredNodes(lookupParents, nodeKey.instanceKeys);
+  if (filteredNodes.length === 0) {
+    return filterTargets;
+  }
+
+  filteredNodes.forEach((filteredNode) => collectFilterTargets(filterTargets, filteredNode));
+  return filterTargets;
+}
+
+function findMatchingFilteredNodes(lookupParents: Array<{ children?: Map<Id64String, FilteredTreeNode> }>, keys: InstanceKey[]) {
+  return lookupParents
+    .flatMap((lookup) => keys.map((key) => lookup.children?.get(key.id)))
+    .filter((lookupNode): lookupNode is FilteredTreeNode => lookupNode !== undefined);
+}
+
+function collectFilterTargets(filterTargets: FilterTargets, node: FilteredTreeNode) {
+  if (node.isFilterTarget) {
+    addTarget(filterTargets, node);
+    return;
+  }
+
+  if (node.type === "element") {
+    // need to add parent ids as filter target will be an element
+    addTarget(filterTargets, node);
+  }
+
+  if (!node.children) {
+    return;
+  }
+
+  for (const child of node.children.values()) {
+    collectFilterTargets(filterTargets, child);
+  }
```

Comment:
In previous implementation if filter target is `element` it was adding it's `element` ancestors also https://github.com/iTwin/viewer-components-react/blob/485ec9d4bd818fb9ba6aab08af56287c18f41cb7/packages/itwin/tree-widget/src/components/trees/models-tree/internal/ModelsTreeVisibilityHandler.ts#L523

So this does the same because we reach an `element` that is not a filter target it means that it's children is filter target.

Reply:
Okay, seems like I was confused by the function name `collectFilterTargets`. Can we rename it to something more meaningful, e.g. `collectVisibilityChangeTargets` or something like that?

Reply:
Renamed